File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    /*
    use sliding window left =0 and right = 1
    iterate the loop until right pointer exceeds size of input array
    if it is a profitable transaction then calculate profit and update maxprofit
    otherwise move the left pointer to right as we have got a day where left is smaller than prev value
    increment right everytime
    */
    int maxProfit(vector<int>& prices) {
        int left = 0;
        int right = 1;
        int maxProfit = 0;
        
        while(right < prices.size()){
            if(prices[left] < prices[right]){
                int profit = prices[right] - prices[left];
                maxProfit = max(maxProfit,profit);              
            }
            else{
                left = right;
            }
            right++;            
        }
       return maxProfit; 
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int mini = Integer.MAX_VALUE;
        int profit = 0;
        for(int ele: prices){
            mini = Math.min(mini,ele);
            int currProfit = ele - mini;
            profit = Math.max(profit, currProfit);
        }
        return profit;
    }
}